Ticket3585: Autoscales log plot for a short time when plot opened

adamb56789 opened the pull request:
### Description of work

Set auto-scale to true by default so it scales the plot when it is first opened, then added an event listener to turn auto-scale back off after 10 data points are added.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/3585

### Acceptance criteria

- [x] Auto-scale works a single time when a new graph is opened.

### Documentation

https://github.com/ISISComputingGroup/ibex_user_manual/wiki/Plot-a-Block-Graph
Made some changes to clarify some of the things mentioned in the ticket.

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

DominicOram commented:
This works as described but has a couple of issues. I occasionally now get the following exception:
```
Exception in thread "ForkJoinPool-1-worker-0" org.eclipse.swt.SWTException: Invalid thread access
	at org.eclipse.swt.SWT.error(SWT.java:4552)
	at org.eclipse.swt.SWT.error(SWT.java:4467)
	at org.eclipse.swt.SWT.error(SWT.java:4438)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:448)
	at org.eclipse.swt.widgets.Widget.checkWidget(Widget.java:366)
	at org.eclipse.swt.widgets.Table.deselectAll(Table.java:1944)
	at org.eclipse.jface.viewers.TableViewer.doDeselectAll(TableViewer.java:300)
	at org.eclipse.jface.viewers.AbstractTableViewer.setSelectionToWidget(AbstractTableViewer.java:830)
	at org.eclipse.jface.viewers.StructuredViewer.setSelectionToWidget(StructuredViewer.java:1760)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1453)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1407)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1507)
	at org.eclipse.jface.viewers.ColumnViewer.refresh(ColumnViewer.java:523)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1466)
	at org.csstudio.trends.databrowser2.propsheet.TraceTableHandler$1.changedAxis(TraceTableHandler.java:78)
	at org.csstudio.trends.databrowser2.model.Model.fireAxisChangedEvent(Model.java:936)
	at org.csstudio.trends.databrowser2.model.AxisConfig.fireAxisChangeEvent(AxisConfig.java:305)
	at org.csstudio.trends.databrowser2.model.AxisConfig.setAutoScale(AxisConfig.java:285)
	at uk.ac.stfc.isis.ibex.ui.logplotter.LogPlotterHistoryPresenter$1.selectedSamplesChanged(LogPlotterHistoryPresenter.java:128)
	at org.csstudio.trends.databrowser2.model.Model.fireSelectedSamplesChanged(Model.java:975)
	at org.csstudio.trends.databrowser2.ui.Controller$3.selectedSamplesChanged(Controller.java:386)
	at org.csstudio.trends.databrowser2.ui.ModelBasedPlot$2.lambda$3(ModelBasedPlot.java:135)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.csstudio.trends.databrowser2.ui.ModelBasedPlot$2.changedCursors(ModelBasedPlot.java:135)
	at org.csstudio.swt.rtplot.internal.Plot.fireCursorsChanged(Plot.java:1296)
	at org.csstudio.swt.rtplot.internal.Plot.updateCursors(Plot.java:1119)
	at org.csstudio.swt.rtplot.internal.PlotProcessor.lambda$1(PlotProcessor.java:297)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```
I can somewhat reliably reproduce it by adding two traces to the graph and getting them to share an axis then add another 4~5 traces to the graph.

It would probably be useful to add a comment to the code to say what the listener is doing and why.

This doesn't address the 2nd issue on the original ticket about linking the axes. Can you either address this (maybe by adding the option in the block context menu for which axis to plot on) or pull it into a separate ticket.

adamb56789 commented:
I added a comment about the listener and a new option in the menu to add to an existing axis (and updated the [wiki](https://github.com/ISISComputingGroup/ibex_user_manual/wiki/Plot-a-Block-Graph)).
I encountered the same exception as above during testing, though it did not take that many traces for it to occur. One time I created a new plot and got the error immediately.

adamb56789 commented:
I've followed the internet's instructions and put the offending line in an asyncExec.
```java
Display.getDefault().asyncExec(new Runnable() {
     public void run() {
        axis.setAutoScale(false);
    }
});
```
I made several attempts to reproduce the bug after the change, and I didn't get it to happen.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.blocks.presentation/src/uk/ac/stfc/isis/ibex/ui/blocks/presentation/AxisList.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class AxisList {
+    private String plotName;
+    private Stream<String> axisNames;
```
I've changed it.

DominicOram commented:
This works well and the exception I was getting previously is now gone. However, there are a few bits of functionality I would change:
* The context menu is getting very long now can we instead have submenus so that I can click on `Display Block History` -> `Add to XYZ plot` -> `Add to new axis`
* The colour of the axes is now getting confusing. Maybe if you add a second trace to an axis it's colour changes to a neutral black? This is a nitpick though.

I also think the code could be a little tidier:
* The LogPlotterHistoryPresenter class now has quite a bit of duplicated code, specifically between `addPVToEditor` and `addPVToAxis` and between `addToDisplay` and `addToAxis`. Can you try and refactor this to reduce the duplication? Maybe have a look at [Optional](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html), which might help.
* I think you could replace the `AxisList` class with a HashMap.
* Rather than have the two methods `getDataBrowserTitles` and `getAxisTitles` you can probably just have one as all the information for `getDataBrowserTitles` is in `getAxisTitles`
* I've added a couple of minor points in the code too.

DominicOram commented on base/uk.ac.stfc.isis.ibex.ui.logplotter/src/uk/ac/stfc/isis/ibex/ui/logplotter/LogPlotterHistoryPresenter.java:
```diff
@@ -113,13 +126,79 @@ private void addPVToEditor(String pvAddress, final String displayName, DataBrows
+            item.useDefaultArchiveDataSources();
+
+            // Extract the axis to add to from the list of axes
+            List<AxisConfig> axes = StreamSupport.stream(model.getAxes().spliterator(), false).filter(a -> a.getName() == axisName).collect(Collectors.toList());
```
You only want one axis, take a look at the `findFirst` method

DominicOram commented on base/uk.ac.stfc.isis.ibex.ui.logplotter/src/uk/ac/stfc/isis/ibex/ui/logplotter/LogPlotterHistoryPresenter.java:
```diff
@@ -113,13 +126,79 @@ private void addPVToEditor(String pvAddress, final String displayName, DataBrows
+	                if (dataCount < DATA_COUNT_LIMIT) {
+	                    dataCount++;
+	                } else if (autoscale) {
+	                    Display.getDefault().asyncExec(new Runnable() {
```
There's some slightly leaner syntax you can use here: `Display.getDefault().asyncExec(() -> axis.setAutoScale(false));`

adamb56789 commented:
I've made all of the suggested changes.
